### DISCUSSION TITLE
SDK 1.9.0 notice type 29- OPT-301-Tenderer-MainCont mandatory but not OPT-301-Tenderer-SubCont

### DISCUSSION THREAD

**Discussion by luispozoWS** [OP-TED/eForms-SDK#917] [Q&A] [answered]
Hello,

OPT-301-Tenderer-MainCont is mandatory when OPT-300-Tenderer is filled (BR-OPT-00301-0187)
Why OPT-301-Tenderer-SubCont isn't mandatory ? (as not filling it but filling OPT-300-Tenderer and OPT-301-Tenderer-MainCont creates an XSD error)
Why is GR-Tenderer-Main-Contractor-Id a sub-group of GR-Subcontractor ? Therefore the "normal" behavior for the user is not filling OPT-300-Tenderer (declaring a tenderer) if a subcontractor doesn't exists ?
Is this going to change in any SDK version ? 

![image](https://github.com/OP-TED/eForms-SDK/assets/158502029/1caf809c-7d93-4c9f-9eb2-2046fddeaac0)

Thanks
Best Regards

**Comment by YvesJo** [comment, marked as answer]
The context for the rule _BR-OPT-00301-0187_ is the node _ND-SubContractor_ hence only applies when _efac:Subcontractor_ exists.
For _efac:SubContractor_ to exist, its child _cbc:ID_ (_OPT-301-Tenderer-SubCont_) shall also exist (required by the XSD and no empty nodes allowed).
It is not possible to mark _OPT-301-Tenderer-SubCont_ as mandatory as there may be no subcontractors. To identify "_the Main Tenderers a SubContractor would be providing services to_" (_OPT-301-Tenderer-MainCont_), the user shall first identify the SubContractor (_OPT-301-Tenderer-SubCont_).
As the validation process starts with the lexical validation (XSD conformance) before any schematron rule is applied, when the XML does not comply with the schema, an XSD error is triggered.
To illustrate my answer to the question on GR-Tenderer-Main-Contractor-Id, I've added the diagram below where:
- information about the (Main) Tenderers is framed in red,
- information about the SubContractors is framed in green.

The display group _GR-Tenderer-Main-Contractor-Id_ is inside _GR-Subcontractor_ because it lists "_the Main Tenderers a SubContractor would be providing services to_" (_OPT-301-Tenderer-MainCont_); its associated label is also "_Subcontractor taker_".
The rule _BR-OPT-00301-0187_ also prevents the specification of _OPT-301-Tenderer-MainCont_ so long the Main Tenderers (_OPT-300-Tenderer_) aren't themselves identified.

![image](https://github.com/OP-TED/eForms-SDK/assets/42802091/6d96d0cc-0ec8-4264-ad3e-c2b2971ed31a)

**Comment by luispozoWS** [comment]
Thanks for your answer @YvesJo.

Our developpers state that "efac:subcontractor" is a mandatory element in the XML (when efac:TenderingParty exists) as it contains the field "OPT-301-Tenderer-MainCont" that, for them, is mandatory when OPT-300-Tenderer is filled. Do you agree on that statement ?

"efac:subcontractor" should exist by default when "efac: TenderingParty" exists in the XML ? Or not?
Or should it be conditionned to filling fields OPT-300-Tenderer and OPT-301-Tenderer-SubCont ?
Schematron rule BR-OPT-00301-0187 is meant to appear only when efac:subcontractor exists ?

**Reply by YvesJo** [reply]
Hi,

> Our developpers state that "efac:subcontractor" is a mandatory element in the XML (when efac:TenderingParty exists) as it contains the field "OPT-301-Tenderer-MainCont" that, for them, is mandatory when OPT-300-Tenderer is filled.

This is not actually true, there is a context {ND-SubContractor} associated to the rule for the rule to get triggered. May be consideration of the context is a challenge for the UI, however CVS will not issue an error when the node is not present.

This node is associated to the aggregate element efac:SubContractor which may not exist (no empty aggregate element allowed) if neither OPT-301-Tenderer-MainCont nor OPT-301-Tenderer-SubCont exist; efac:SubContractor should only exist when it has some content.

With SDK 1.12, the interdependency has been rewritten to get closer to the business logic

> Schematron rule BR-OPT-00301-0187 is meant to appear only when efac:subcontractor exists?

Yes, this is exact.
KR
